Implement allowed hosts filter route modifiers

easel opened the pull request:
# Pull Request Checklist

* [X] Have you read [How to write the perfect pull request](https://github.com/blog/1943-how-to-write-the-perfect-pull-request)?
* [X] Have you read through the [contributor guidelines](https://www.playframework.com/contributing)?
* [X] Have you signed the [Lightbend CLA](https://www.lightbend.com/contribute/cla)?
* [X] Have you referenced any issues you're fixing using [commit message keywords](https://help.github.com/articles/closing-issues-using-keywords/)?
* [X] Have you added copyright headers to new files?
* [X] Have you checked that both Scala and Java APIs are updated?
* [X] Have you updated the documentation for both Scala and Java sections?
* [X] Have you added tests for any changed functionality?

## Purpose
In some cases requests cannot always be validated against a fixed list of Allowed Hosts. Specifically, autoscaling groups using AWS application load balancers require http host checks, and will always use the dynamically assigned IP address of the instance as the host. This PR allows the allowed hosts filter to be applied to or filtered out of specific routes (such as health check urls) if desired. 

## Background Context
This PR adds a the ability to whitelist or blacklist route modifiers from the AllowedHostsFilter. The basic premise mirrors that of the CSRFilter.

marcospereira commented on framework/src/play-filters-helpers/src/main/scala/play/filters/hosts/AllowedHostsFilter.scala:
```diff
@@ -65,21 +64,38 @@ private[hosts] case class HostMatcher(pattern: String) {
 }
 
-case class AllowedHostsConfig(allowed: Seq[String]) {
+case class AllowedHostsConfig(allowed: Seq[String],
```
This breaks binary compatibility. If you plan to backport this PR, it could be good to have another constructor emulating the existing one (only receiving `Seq[String]` with allowed hosts and falling back other fields to default values). You will also need to provide a `copy` method.

marcospereira replied:
Binary compatibility problems reported here: https://travis-ci.org/playframework/playframework/jobs/420522793#L962-L967

easel replied:
I *think* the latest push properly addresses the binary compatibility issues with the AllowedHostsConfig.

marcospereira commented on framework/src/play-filters-helpers/src/main/scala/play/filters/hosts/AllowedHostsFilter.scala:
```diff
@@ -65,21 +64,38 @@ private[hosts] case class HostMatcher(pattern: String) {
+  def withWhiteListRouteModifier(routeModifiers: Seq[String]): AllowedHostsConfig = copy(whiteListRouteModifiers = routeModifiers)
+  def withBlackListRouteModifier(routeModifiers: Seq[String]): AllowedHostsConfig = copy(blackListRouteModifiers = routeModifiers)
+
+  @inline def checkRouteModifiers(rh: RequestHeader): Boolean = {
```
Please, add some tests for this.

gmethvin replied:
Does this need to be public?

marcospereira replied:
I don't think so. To be clear, I'm mostly worried about having tests for when there are modifiers in the whitelist and blacklist (when when these are empty too).

easel replied:
It's now part of the AllowedHostsFilter fromConfiguration constructor.

gmethvin commented on framework/src/play-filters-helpers/src/main/scala/play/filters/hosts/AllowedHostsFilter.scala:
```diff
@@ -65,21 +64,38 @@ private[hosts] case class HostMatcher(pattern: String) {
-case class AllowedHostsConfig(allowed: Seq[String]) {
+case class AllowedHostsConfig(allowed: Seq[String],
+                              whiteListRouteModifiers: Seq[String],
+                              blackListRouteModifiers: Seq[String]) {
```
It might be more versatile to add something like:

```scala
shouldProtect: RequestHeader => Boolean
```

This is what we do in the CSRF filter. This way users can provide the config manually and filter based on any criteria they want. It also means we can later add configuration in `application.conf` without breaking binary compatibility.

easel replied:
Good idea. Done.

easel commented:
Thanks for the feedback @marcospereira and @gmethvin, it is exactly what I was hoping for. I think all the requests are reasonable, and the pointers to tests and docs will be very helpful. I'll submit an updated revision ASAP.

marcospereira commented on framework/src/play-filters-helpers/src/main/scala/play/filters/hosts/AllowedHostsFilter.scala:
```diff
@@ -4,15 +4,14 @@
 import play.api.libs.streams.Accumulator
-import play.api.mvc.{ EssentialAction, EssentialFilter }
-import play.core.j.{ JavaContextComponents, JavaHttpErrorHandlerAdapter }
+import play.api.mvc.{EssentialAction, EssentialFilter, RequestHeader}
```
Small thing I noticed now. There is a code format issue here. Running `sbt scalariformFormat test:scalariformFormat` inside `framework` folder and pushing the changes should fix it.

easel commented:
Ok @marcospereira @gmethvin I think I've addressed all your feedback and have added tests and documentation. The PR has been updated with a newly squashed commit and rebased against latest master. Take a look and let me know what you think. Thanks!

PromanSEW commented on framework/src/play-filters-helpers/src/main/scala/play/filters/hosts/AllowedHostsFilter.scala:
```diff
@@ -65,21 +65,46 @@ private[hosts] case class HostMatcher(pattern: String) {
 
   import scala.collection.JavaConverters._
-  def withHostPatterns(hosts: java.util.List[String]): AllowedHostsConfig = copy(hosts.asScala.toSeq)
+  import java.{ util => ju }
```
Unused import

easel replied:
It is part of the “withShouldProtect” java compat signature below.

easel replied:
I went ahead and scrubbed out the import in the name of consistency since the extant withHostPattens uses java.util directly. Thanks!

easel commented:
Pushed another couple updates cleaning up import consistency and binary compatibility. Hopefully Travis goes green on this one.

gmethvin commented on framework/src/play-filters-helpers/src/main/scala/play/filters/hosts/AllowedHostsFilter.scala:
```diff
@@ -65,22 +65,58 @@ private[hosts] case class HostMatcher(pattern: String) {
+  def copy(allowed: Seq[String], shouldProtect: RequestHeader => Boolean): AllowedHostsConfig =
+    new AllowedHostsConfig(allowed = allowed, shouldProtect = shouldProtect)
+
+  def copy(shouldProtect: RequestHeader => Boolean): AllowedHostsConfig =
```
Is this one needed for binary compatibility?

easel replied:
No, I don't think so, but it's needed to maintain consistency with the defaults you'd expect from a case class and is used by the java api -- see `withShouldProtect`.

gmethvin commented on framework/src/play-filters-helpers/src/main/scala/play/filters/hosts/AllowedHostsFilter.scala:
```diff
@@ -65,22 +65,58 @@ private[hosts] case class HostMatcher(pattern: String) {
+  def copy(shouldProtect: RequestHeader => Boolean): AllowedHostsConfig =
+    copy(allowed = allowed, shouldProtect = shouldProtect)
+
+  def copy(allowed: Seq[String] = allowed): AllowedHostsConfig =
```
If we need `copy` methods for bincompat, I'd make them `private[hosts]` so these are visible at runtime but not to the Scala compiler. This way all new code will be compiled with the new version.

easel replied:
Ok, done.

gmethvin commented on documentation/manual/working/commonGuide/filters/AllowedHostsFilter.md:
````diff
@@ -34,6 +34,40 @@ play.filters.hosts {
 }
 ```
 
+## WhiteListing
````
"Whitelisting" (I prefer not to use CamelCase in normal English writing.)

We can keep the configuration `whiteList` since that's consistent with what's used elsewhere in Play.

gmethvin replied:
I also noticed "white-list" and "blackList" below. Let's be consistent and use "whitelist" and "blacklist" when referring to them.

Perhaps the whitelisting/blacklisting headers can be under a common header "Applying to routes based on route modifiers" (or something like that)

easel replied:
Makes sense. I was trying to figure out what to do to maintain consistency with the config values.

gmethvin commented:
> binary compatibility

Did you intend to backport to 2.6?

easel commented:
Updated per review comments from @gmethvin. I'm ambivalent about backporting to 2.6. Since it's backwards compatible I guess it doesn't hurt?

gmethvin commented on framework/src/play-filters-helpers/src/main/scala/play/filters/hosts/AllowedHostsFilter.scala:
```diff
@@ -65,22 +65,58 @@ private[hosts] case class HostMatcher(pattern: String) {
     this(Seq.empty)
   }
 
+  private[hosts] def copy(allowed: Seq[String], shouldProtect: RequestHeader => Boolean): AllowedHostsConfig =
```
isn't this the same as the automatically generated copy method?

easel replied:
It is, but I believe by defining any copy method the automatically generated one gets dropped.

gmethvin commented on framework/src/play-filters-helpers/src/main/scala/play/filters/hosts/AllowedHostsFilter.scala:
```diff
@@ -65,22 +65,58 @@ private[hosts] case class HostMatcher(pattern: String) {
+  private[hosts] def copy(allowed: Seq[String], shouldProtect: RequestHeader => Boolean): AllowedHostsConfig =
+    new AllowedHostsConfig(allowed = allowed, shouldProtect = shouldProtect)
+
+  private[hosts] def copy(shouldProtect: RequestHeader => Boolean): AllowedHostsConfig =
```
where is this one used?

easel replied:
I'm going on the output from Mima.

gmethvin commented on framework/src/play-filters-helpers/src/main/scala/play/filters/hosts/AllowedHostsFilter.scala:
```diff
@@ -65,22 +65,58 @@ private[hosts] case class HostMatcher(pattern: String) {
 
-case class AllowedHostsConfig(allowed: Seq[String]) {
+case class AllowedHostsConfig(allowed: Seq[String], shouldProtect: RequestHeader => Boolean) {
+  def this(allowed: Seq[String]) {
```
I think all these backwards compatibility changes should be added in a separate 2.6.x PR.

We'll probably want to add this and the alternate `apply` method as `private[hosts]` and make `shouldProtect` have a default value.

easel replied:
Ok, so how do we deal with the travis mima failures, just ignore them? Given that, for this PR, should I just drop everything that isn't strictly required for the tests to pass?

easel replied:
I went ahead and unwound the bincompat stuff in a separate commit. If that's what you had in mind I'll squash and we can go from there.

marcospereira replied:
> Ok, so how do we deal with the travis mima failures, just ignore them? 

You add filters for them:

https://github.com/playframework/playframework/blob/e4485b35314a810168673042f610158566f9df5a/framework/project/BuildSettings.scala#L181

For Play 2.6.x, the build should pass without adding filters, in other words, binary compatibility should be kept. But for master, it is okay to break it and the filters are good so that we can know what is breaking and why.

easel commented:
Ok, filters added, squashed, and rebased, and looks like Jenkins went green.

marcospereira commented:
Thank you, @easel!

easel commented:
Thanks for the feedback and merge @marcospereira @gmethvin
